training/CHAPEL_TUTORIAL.md

```diff
@@ -0,0 +1,846 @@
+  // for loop iteration `j`, `vi` will refer to the `j`th element of `Ai`
+}
+```
+Since all `Ai` have compatible domains, we know all `vi` will be local to one another on any given loop iteration.
+
+<a id="ex14"></a>
+#### Example 14: Simple Zippered Iteration
+Let's start with a simple example involving zippered iteration:
+```Chapel
+var A: [1..5] real;
+for (a, i, j) in zip(A, 1.., [3, 0, 1, 2, 4]) {
+  a = i**j;
+}
+writeln(A);
+```
+```console
+1.0 1.0 3.0 16.0 625.0
+```
+
+<a id="ex15"></a>
+#### Example 15: Zippered Iteration in Arkouda
+Now let's look at an example based on [`getLengths`](https://github.com/Bears-R-Us/arkouda/blob/b86adeb843275b0b86553534ede632acef4d15e2/src/SegmentedString.chpl#L388-L406) in `SegmentedString.chpl`.
+
+First we need some context. An Arkouda `Strings` (aka SegString) is an array of variable length strings.
+```python
+>>> s = ak.array([f"{'s'*(3-i)}{i}" for i in range(3)])
+>>> s
+array(['sss0', 'ss1', 's2'])
+>>> type(s)
+arkouda.strings.Strings
+```
+
+On the server, a SegString is made up of two distributed array components:
+- values: `uint(8)`
+  - the flattened array of bytes in all the strings (null byte delimited)
+- offsets: `int`
+  - the start index of each individual string in the `values`
+
+For the sake of simplicity, we'll treat values as a string array. So for our `s` above, these components look something like
+```
+values = ['\x00', 's', 's', 's', '0', '\x00', 's', 's', '1', '\x00', 's', '2']
+offsets = [0, 5, 9]
+```
+For `getLengths`, we want to calculate the length of each individual string including the null terminator.
+
+```chapel
+const values: [0..#12] string = ['\x00', 's', 's', 's', '0', '\x00', 's', 's', '1', '\x00', 's', '2'];
+const offsets = [0, 5, 9];
+const size = 3;
+
+/* Return lengths of all strings, including null terminator. */
+proc getLengths() {
+  // initialize lengths with the same domain as offsets
+  var lengths: [offsets.domain] int;
+  if size == 0 {
+    // if the segstring is empty, the lengths are also empty
+    return lengths;
+  }
+  // save off the last index of offsets
+  const high = offsets.domain.high;
+  forall (i, o, l) in zip(offsets.domain, offsets, lengths) {
+    if i == high {
+      // the last string
+      // len = values.size - start position of this string
+      l = values.size - o;
+    }
+    else {
+      // not the last string
+      // len = start poistion of next string - start position of this string
+      l = offsets[i+1] - o;
+    }
+  }
+  return lengths;
+}
+
+writeln(getLengths());
+```
+```console
+5 4 3
+```
+Nice! We'll use `getLengths`, `values`, and `offsets` in the next section as well.
+
+<a id="aggregation"></a>
+### Aggregation
+
+[Aggregation](https://chapel-lang.org/docs/modules/packages/CopyAggregation.html#copyaggregation) is used to copy a
+local value into a remote array or copy a value from a remote array into a local variable. This provides a significant speed up when doing batch assignments.
```

Comment:
Since we are only demoing the DstAggegrator, I think with would be useful to  to update `is used to copy a
local value into a remote array or copy a value from a remote array into a local variable` to `is used to copy a
local value into a remote array (DstAggregator) or copy a value from a remote array into a local variable (SrcAggegator)

Reply:
Another thing here is maybe we should say values plural, rather than singular, since aggregation is typically used for portions of arrays and not just single elements? (I feel like this is implied by the note about batch assignments)

Reply:
great suggestions! I'll be sure to update this!